### PR TITLE
Deal with existing state when starting a memory service

### PR DESCRIPTION
**Pull request by vanstee**
* Remonitor or delete related data for existing processes on init
* Periodically check all keys and remove anything that could have snuck in between process restarts

### CODE REVIEW

File: lib/cog/command/service/memory.ex [hunk 1]
```diff
@@ -59,100 +61,139 @@ defmodule Cog.Command.Service.Memory do
...
+
+    {:noreply, state}
+  end
+
+  defp account_for_existing_processes(monitor_table, memory_table) do
+    ets_iterate(monitor_table, fn pid, token ->
+      case Process.alive?(pid) do
+        true ->
+          Logger.debug("Remonitoring #{inspect pid} for token #{inspect token}")
+          :erlang.monitor(:process, pid)
```
Comment (christophermaier) [inline lib/cog/command/service/memory.ex]: Don't you need to keep track of this monitor?

File: lib/cog/command/service/memory.ex [hunk 2]
```diff
@@ -59,100 +61,136 @@ defmodule Cog.Command.Service.Memory do
...
+        false ->
+          cleanup_process(monitor_table, memory_table, pid, token)
+      end
+    end)
+  end
+
+  defp schedule_dead_process_cleanup(interval) do
+    Logger.info ("Scheduling dead process cleanup for #{round(interval / 1000)} from now")
```
Comment (christophermaier) [inline lib/cog/command/service/memory.ex]: Should add the units in the log message

File: lib/cog/command/service/memory.ex (hunk 1 above)
Comment (vanstee) [inline lib/cog/command/service/memory.ex]: I don't think so. The pid is already in the ets table and we're calling this before init returns so all pids in the ets table should be monitored and we can use that to make sure we're not double monitoring.

Comment (christophermaier) [inline lib/cog/command/service/memory.ex]: Ah, yeah... I just got to that part below. 👍

File: test/cog/command/service/memory_test.exs [hunk 3]
```diff
@@ -6,6 +6,30 @@ defmodule Cog.Command.Service.MemoryTest do
...
+  end
+
+  setup do
+    memory_table  = :ets.new(:test_memory_table,  [:public])
+    monitor_table = :ets.new(:test_monitor_table, [:public])
+
+    {:ok, pid} = Memory.start_link(memory_table, monitor_table)
+
+    on_exit(fn ->
+      Process.exit(pid, :kill)
```
Comment (christophermaier) [inline test/cog/command/service/memory_test.exs]: Is this necessary if you've `start_linked` the process?

**Comment by christophermaier** [comment]
👍

File: test/cog/command/service/memory_test.exs (hunk 3 above)
Comment (vanstee) [inline test/cog/command/service/memory_test.exs]: Ah you're right.

**Pull request merged**